pybart/new_matcher.py

```diff
@@ -0,0 +1,330 @@
+            #   because it is a negative search (that is non of the labels should match)
+            for actual_label in actual_labels:
+                if (is_regex and re.match(constraint.value[1:-1], actual_label)) or (constraint.value == actual_label):
+                    return False, set()
+        elif isinstance(constraint, HasLabelFromList):
+            current_successfully_matched = set()
+            # at least one of the constraint strings should match, so return False only if none of them did.
+            for value_option in constraint.value:
+                # check if a regex or exact match is required
+                is_regex = value_option.startswith('/') and value_option.endswith('/')
+
+                # for each edged label, check if the label matches the constraint, and store it if it does,
+                #   because it is a positive search (that is at least one label should match)
+                for actual_label in actual_labels:
+                    if (is_regex and re.match(value_option[1:-1], actual_label)) or (value_option == actual_label):
+                        # store the matched label
+                        current_successfully_matched.add(actual_label)
+            # this means no match, so fail
+            if not current_successfully_matched:
+                return False, set()
+            
+            # concat the captured labels of the current constraint with previous ones, as there can be a few positive constraints.
+            successfully_matched.update(current_successfully_matched)
+    
+    return True, successfully_matched
+
+
+class SentenceMatch:
+    def __init__(self, name2index: Mapping[str, int], indices2label: Mapping[Tuple[int, int], Set[str]]):
+        self.name2index = name2index
+        self.indices2label = indices2label
+    
+    # return the index of the specific matched token in the sentence according to its name
+    def token(self, name: str) -> int:
+        return self.name2index.get(name, default=-1)  # TODO - maybe raise instead of returning -1? it means typo in coding
+    
+    # return the set of captured labels between the two tokens, given their indices
+    def edge(self, t1: int, t2: int) -> Set[str]:
+        return self.indices2label.get((t1, t2), default=None)  # TODO - is it legitimate to return None here?
+
+
+class GlobalMatcher:
+    def __init__(self, constraint: Full):
+        self.constraint = constraint
+        self.captured_labels = defaultdict(set)
+        # list of token ids that don't require a capture
+        self.dont_capture_names = [token.id for token in constraint.tokens if not token.capture]
+    
+    # filter a single match group according to non-token constraints
+    def _filter(self, match: Mapping[str, int], sentence) -> bool:
+        # check that the distance between the tokens is not more than or exactly as required
+        for distance in self.constraint.distances:
```

Comment:
split into separate functions, for edges, distances, concats, ... and one function that calls them one after the other. it will also make things much more testable.